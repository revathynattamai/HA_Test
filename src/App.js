import React, { useState } from "react";
import Card from "@material-ui/core/Card";
import CardActionArea from "@material-ui/core/CardActionArea";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import Typography from "@material-ui/core/Typography";
import Pagination from "@material-ui/lab/Pagination";

import { useData } from "./hooks/useData";
import { useStyles } from "./styles";

export default function App() {
  const [page, setPage] = useState(1);
  const data = useData(page);
  const classes = useStyles();

  const onPageChange = (ev, value) => {
    setPage(value);
  };

  if (!data) return <div>Loading..</div>;

  const {
    info,
    results,
    error
  } = data;

  if (error) return <div>{error}</div>;

  return (
    <div className={classes.wrapper}>
      <div className={classes.content}>
        {results?.map((character) => (
          <Character key={character.id} {...character} />
        ))}
      </div>
      <Pages value={page} total={info?.pages} onPageChange={onPageChange} />
    </div>
  );
}

const Character = ({ name, image, species, location, origin }) => {
  const classes = useStyles();

  return (
    <Card className={classes.root}>
      <CardActionArea>
        <CardMedia className={classes.media} image={image} title={name} />
        <CardContent>
          <Typography gutterBottom variant="h5" component="h2">
            {name}
          </Typography>
          <Typography variant="subtitle1">{species}</Typography>
          <Typography variant="body2">{location?.name}</Typography>
          <Typography variant="body2">{origin?.name}</Typography>
        </CardContent>
      </CardActionArea>
    </Card>
  );
};

const Pages = ({ value, total, onPageChange }) => {
  const classes = useStyles();

  return (
    <Pagination
      id='pagination'
      page={value}
      count={total}
      onChange={onPageChange}
      className={classes.pagination}
    />
  );
};
