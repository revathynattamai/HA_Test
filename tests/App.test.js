import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

import App from "../src/App";

let container = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it("renders app successfully", async () => {
  const fakeData = {
    info: {
      count: 671,
      pages: 34,
    },
    results: [
      {
        id: 1,
        name: "Rick Sanchez",
        species: "Human",
        origin: {
          name: "Earth (C-137)",
        },
        image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      }
    ]
  };

  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeData)
    })
  );

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<App />, container);
  });

  expect(container).toMatchSnapshot();

  // remove the mock to ensure tests are completely isolated
  global.fetch.mockClear();
  delete global.fetch;
});

it("renders new page successfully", async () => {
  const fakeData = {
    info: {
      count: 671,
      pages: 34,
    },
    results: [
      {
        id: 1,
        name: "Rick Sanchez",
        species: "Human",
        origin: {
          name: "Earth (C-137)",
        },
        image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      }
    ]
  };

  const fakeData2 = {
    info: {
      count: 671,
      pages: 34,
    },
    results: [
      {
        id: 1,
        name: "Rick Sanchez 2",
        species: "Human",
        origin: {
          name: "Earth (C-137)",
        },
        image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
      }
    ]
  };

  global.fetch = jest.fn()
    .mockImplementationOnce(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData)
      }))
    .mockImplementationOnce(() =>
      Promise.resolve({
        json: () => Promise.resolve(fakeData2)
      })
    );

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<App />, container);
  });

  await act(async () => {
    container.querySelector("#pagination > ul > li:nth-child(3) > button").click();
  });

  expect(container).toMatchSnapshot();

  // remove the mock to ensure tests are completely isolated
  global.fetch.mockClear();
  delete global.fetch;
});

it("renders app successfully with uneven data", async () => {
  const fakeData = {};

  global.fetch = jest.fn().mockImplementation(() =>
    Promise.resolve({
      json: () => Promise.resolve(fakeData)
    })
  );

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<App />, container);
  });

  expect(container).toMatchSnapshot();

  // remove the mock to ensure tests are completely isolated
  global.fetch.mockClear();
  delete global.fetch;
});

it("renders error", async () => {
  global.fetch = jest.fn().mockImplementation(() =>
    Promise.reject("Error")
);

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(<App />, container);
  });

  expect(container).toMatchSnapshot();

  // remove the mock to ensure tests are completely isolated
  global.fetch.mockClear();
  delete global.fetch;
});
