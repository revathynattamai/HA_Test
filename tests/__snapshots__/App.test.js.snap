// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders app successfully 1`] = `
<div>
  <div
    class="makeStyles-wrapper-1"
  >
    <div
      class="makeStyles-content-2"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-3 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-media-4"
            style="background-image: url(https://rickandmortyapi.com/api/character/avatar/1.jpeg);"
            title="Rick Sanchez"
          />
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              Rick Sanchez
            </h2>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
            >
              Human
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body2"
            />
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Earth (C-137)
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root makeStyles-pagination-5"
      id="pagination"
    >
      <ul
        class="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 1"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-selected"
            tabindex="0"
            type="button"
          >
            1
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 2"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            2
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 3"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            3
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 4"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            4
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 5"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            5
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <div
            class="MuiPaginationItem-root MuiPaginationItem-ellipsis"
          >
            â€¦
          </div>
        </li>
        <li>
          <button
            aria-label="Go to page 34"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            34
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`renders app successfully with uneven data 1`] = `
<div>
  <div
    class="makeStyles-wrapper-11"
  >
    <div
      class="makeStyles-content-12"
    />
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root makeStyles-pagination-15"
      id="pagination"
    >
      <ul
        class="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 1"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-selected"
            tabindex="0"
            type="button"
          >
            1
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;

exports[`renders error 1`] = `
<div>
  <div>
    Some error occurred, please try again.
  </div>
</div>
`;

exports[`renders new page successfully 1`] = `
<div>
  <div
    class="makeStyles-wrapper-6"
  >
    <div
      class="makeStyles-content-7"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-8 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="MuiCardMedia-root makeStyles-media-9"
            style="background-image: url(https://rickandmortyapi.com/api/character/avatar/1.jpeg);"
            title="Rick Sanchez 2"
          />
          <div
            class="MuiCardContent-root"
          >
            <h2
              class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
            >
              Rick Sanchez 2
            </h2>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
            >
              Human
            </h6>
            <p
              class="MuiTypography-root MuiTypography-body2"
            />
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              Earth (C-137)
            </p>
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <nav
      aria-label="pagination navigation"
      class="MuiPagination-root makeStyles-pagination-10"
      id="pagination"
    >
      <ul
        class="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 1"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            1
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 2"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-selected"
            tabindex="0"
            type="button"
          >
            2
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 3"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            3
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 4"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            4
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 5"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            5
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <div
            class="MuiPaginationItem-root MuiPaginationItem-ellipsis"
          >
            â€¦
          </div>
        </li>
        <li>
          <button
            aria-label="Go to page 34"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            34
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
